app/src/main/java/org/mozilla/rocket/partner/PartnerActivator.java

```diff
@@ -0,0 +1,317 @@
+                    JSONObject activationJsonObject = (JSONObject) activationJsonArray.get(i);
+                    Activation activation = Activation.from(activationJsonObject);
+                    if (activation.matchKeys(activationKeys)) {
+                        found = activation;
+                        break;
+                    }
+                }
+
+                if (found == null) {
+                    //  Activation Not Found
+                    PartnerUtil.log("FetchActivation activation not found, disabled");
+                    partnerActivator.setStatus(Status.Disabled);
+                    return;
+                }
+
+                partnerActivator.activation = found;
+
+                if (found.duration != 0) {
+                    partnerActivator.setSnoozeDuration(found.duration);
+                }
+                if (partnerActivator.inSnooze()) {
+                    partnerActivator.setStatus(Status.Snooze);
+                    PartnerUtil.log("FetchActivation update snoozed");
+                    return;
+                } else {
+                    partnerActivator.setLastCheckedTimestamp(System.currentTimeMillis());
+                }
+
+                ThreadUtils.postToBackgroundThread(new PingActivation(partnerActivator));
+
+            } catch (Exception e) {
+                PartnerUtil.log(e, "FetchActivation Exception");
+            } finally {
+                TrafficStats.clearThreadStatsTag();
+            }
+        }
+    }
+
+    private static class PingActivation extends ActivationJobs implements Runnable {
+
+        PingActivation(PartnerActivator partnerActivator) {
+            super(partnerActivator);
+        }
+
+        @Override
+        public void run() {
+            Activation activation = partnerActivator.activation;
+
+            URL url = null;
+            try {
+                url = new URL(activation.url);
+            } catch (MalformedURLException e) {
+                PartnerUtil.log(e, "PingActivation URL malformed");
+            }
+
+            if (url == null) {
+                PartnerUtil.log("PingActivation URL not found");
+                return;
+            }
+
+            try {
+                TrafficStats.setThreadStatsTag(SOCKET_TAG_PARTNER);
+
+                HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+                urlConnection.setRequestMethod("POST");
+                urlConnection.setDoInput(true);
+                urlConnection.setDoOutput(true);
+
+                urlConnection.setRequestProperty("Content-Type", "application/json");
+
+
+                JSONObject root = new JSONObject();
+                JSONObject fields = new JSONObject();
+                root.put("fields", fields);
+                fields.put("device_id", new JSONObject("{\"stringValue\":\"" + PartnerUtil.getDeviceIdentifier(partnerActivator.context.getContentResolver()) + "\"}"));
```

Comment:
I'm not sure why we're wrapping the child with "{\"stringValue\"...
If we really need to do so, consider writing a function for doing it.